### PR TITLE
Support Union types in `ScalarValue` 

### LINKED ISSUE #8499: Union types not handled when building `AnalysisContext::try_from_column`
### Describe the bug

`AnalysisContext` will try and create a `ScalarValue` when there are no statistics for a given column. Since there is no `ScalarValue` representation for union types this fails. 

The constructor is fallible so this isn't a bug per se, but in other contexts this value is unwrapped and can cause panics during query planning

### To Reproduce

Minimized example of what we have observed causing panics during query planning:
```
    #[test]
    fn test_equivalence_properties_union_type() -> Result<()> {
        let union_type = DataType::Union(
            UnionFields::new(
                vec![0,1],
                vec![
                    Field::new("f1", DataType::Int32, true),
                    Field::new("f2", DataType::Utf8, true),
                ],
            ),
            UnionMode::Sparse,
        );

        let schema = Arc::new(Schema::new(vec![
            Field::new("c1", DataType::Int32, true),
            Field::new("c2", union_type, true),
        ]));

        let exec = FilterExec::try_new(
            binary(
                binary(col("c1", &schema)?, Operator::GtEq, lit(1u32), &schema)?,
                Operator::And,
                binary(col("c1", &schema)?, Operator::LtEq, lit(4u32), &schema)?,
                &schema,
            )?,
            Arc::new(EmptyExec::new(true, schema.clone()))
        )?;

        exec.statistics().unwrap();

        Ok(())
    }
```

Since `FilterExec::equivalence_properties` unwraps the result of `FilterExec::statistics` this just panics. 

### Expected behavior

`FilterExec::equuivalence_properties` is infallible so it shouldn't panic :)

### Additional context

There is probably a way to patch this but seems like we should just add a `ScalarValue::Union` variant.

### PR DESCRIPTION
**Pull request by avantgardnerio**
## Which issue does this PR close?

Closes #8499.
Closes https://github.com/apache/arrow-datafusion/issues/8500

## Rationale for this change

It was decided in the linked issue that implementing the variant was the preferred solution.

## What changes are included in this PR?

Implement the `ScalarValue::Union` variant.

## Are these changes tested?

Via a functional test.

## Are there any user-facing changes?

More queries should work.

### CODE REVIEW

File: datafusion/common/src/scalar/mod.rs [hunk 1]
```diff
@@ -375,6 +379,7 @@ impl PartialEq for ScalarValue {
             (IntervalDayTime(_), _) => false,
             (IntervalMonthDayNano(v1), IntervalMonthDayNano(v2)) => v1.eq(v2),
             (IntervalMonthDayNano(_), _) => false,
+            (Union(_, _, _), _) => false,
```
Comment (thinkharderdev) [inline datafusion/common/src/scalar/mod.rs:382]: We can implement equality here right?

**Review by thinkharderdev — COMMENTED** [review]
One comment but lgtm

**Review by thinkharderdev — APPROVED** [review]

File: datafusion/common/src/scalar/mod.rs [hunk 2]
```diff
@@ -275,6 +277,8 @@ pub enum ScalarValue {
     DurationMicrosecond(Option<i64>),
     /// Duration in nanoseconds
     DurationNanosecond(Option<i64>),
+    /// A nested datatype that can represent slots of differing types. Components:
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:280]: The `"Components: " seem` to be cut off in mid-sentence. Perhaps we could document what each of the 4 fields represents?

File: datafusion/common/src/scalar/mod.rs [hunk 3]
```diff
@@ -500,6 +508,14 @@ impl PartialOrd for ScalarValue {
             (DurationMicrosecond(_), _) => None,
             (DurationNanosecond(v1), DurationNanosecond(v2)) => v1.partial_cmp(v2),
             (DurationNanosecond(_), _) => None,
+            (Union(v1, t1, _m1), Union(v2, t2, _m2)) => {
```
Comment (alamb) [inline datafusion/common/src/scalar/mod.rs:511]: won't this mean that dense unions are compared the same as sparse unions? It seems like we should also compare `m1` and `m2`

**Review by alamb — APPROVED** [review]
Thanks @avantgardnerio  

I think this also closes https://github.com/apache/arrow-datafusion/issues/8500 so added that to the PR description and changed the title to match

File: datafusion/common/src/scalar/mod.rs (hunk 3 above)
Comment (thinkharderdev) [reply datafusion/common/src/scalar/mod.rs:511]: I wonder if we should have the `UnionMode` in here at all? By default calling `to_array_of_size` will generate a sparse union anyway (and creating a dense union from a scalar value wouldn't make sense to begin with....).

Comment (avantgardnerio) [reply datafusion/common/src/scalar/mod.rs:511]: My thought was that if the unions have the same logical data, the physical way they choose to store it is irrelevant to their logical comparison.

File: datafusion/common/src/scalar/mod.rs (hunk 2 above)
Comment (avantgardnerio) [reply datafusion/common/src/scalar/mod.rs:280]: Sorry, this must've gotten wiped in my cherry pick. I restored the rest of the comment.

File: datafusion/common/src/scalar/mod.rs (hunk 3 above)
Comment (alamb) [reply datafusion/common/src/scalar/mod.rs:514]: As long as `ScalarValue` comparison semantics follows the corresponding arrow-rs `UnionArray` comparison semantics, it is good with me

Comment (avantgardnerio) [reply datafusion/common/src/scalar/mod.rs:514]: Looks like that's here: https://github.com/coralogix/arrow-rs/blob/9f20bd8298a52cdde80b0abf19a7582b8b1bfd41/arrow/src/array/equal/union.rs#L120

I'll update to match

Comment (thinkharderdev) [reply datafusion/common/src/scalar/mod.rs]: so we should panic :)

Comment (avantgardnerio) [reply datafusion/common/src/scalar/mod.rs]: I didn't want to go that far...

**Pull request merged**